src/action-providers.ts

```diff
@@ -21,9 +21,73 @@ class RefactoringActionProvider implements vscode.CodeActionProvider {
     const ast = t.parse(document.getText());
     const selection = createSelectionFromVSCode(range);
 
-    return this.refactorings
-      .filter(refactoring => this.canPerform(refactoring, ast, selection))
-      .map(refactoring => this.buildCodeActionFor(refactoring));
+    const applicableRefactorings: RefactoringWithActionProvider[] = [];
+    t.traverseAST(ast, {
+      enter: (path: t.NodePath<any>) => {
+        this.refactorings.forEach(
+          this.checkCanPerform(
+            path,
+            selection,
+            (
+              matchedPath: t.NodePath<any>,
+              refactoring: RefactoringWithActionProvider
+            ) => {
+              if (refactoring.actionProvider.canPerformRefactoringMutator) {
+                applicableRefactorings.push(
+                  refactoring.actionProvider.canPerformRefactoringMutator(
+                    matchedPath,
+                    refactoring
+                  )
+                );
+              } else {
+                applicableRefactorings.push(refactoring);
+              }
```

Comment:
Speaking about "comments to explain why", I think it would worth creating a new ADR document to explain this change too, once we take a decision 👍